### DISCUSSION TITLE
Help with updating esbuild benchmarks for Parcel 2 beta 3

### DISCUSSION THREAD

**Discussion by evanw** [parcel-bundler/parcel#6321] [General] [answered]
Congrats on the Rust rewrite! This is very exciting.

I would like to update my three.js benchmark on https://esbuild.github.io/ to reflect Parcel 2 beta 3. However, I am unable to reproduce the 3x performance improvement advertised on https://v2.parceljs.org/blog/beta3/. The benchmark I'm using can be constructed like this:

```
git clone --depth 1 --branch r108 https://github.com/mrdoob/three.js.git
mkdir src
for i in 1 2 3 4 5 6 7 8 9 10; do cp -r three.js/src "src/copy$$i"; done
for i in 1 2 3 4 5 6 7 8 9 10; do echo "import * as copy$$i from './copy$$i/Three.js'; export {copy$$i}" >> src/entry.js; done
echo 'import * as THREE from "./src/entry.js"; window.THREE = THREE' > entry.parcel2.js
```

Each time I run the benchmark I'm doing this (the cache is cleared because the benchmark measures worst-case build time):

```
rm -fr .parcel-cache
time -p node node_modules/.bin/parcel build entry.parcel2.js
```

Here are the fastest times I could measure on my laptop for each version (MacBook Pro (15-inch, 2019), 2.6 GHz 6-Core Intel Core i7, 16 GB RAM):

| `parcel@2.0.0-beta.2` | `parcel@2.0.0-beta.3.1` | Ratio |
|---|---|---|
| 69 seconds | 39 seconds | 1.8x faster |

I was expecting the build time to be closer to 3x faster but I only observed it being 1.8x faster. Is there anything special that I need to do to enable the SWC backend? I just want to make sure I have the right setup and Parcel command before I update the benchmark.

**Comment by mischnic** [comment, marked as answer]
I'm getting 80 seconds -> 46 seconds on my machine, so 1.7x faster. Maybe @devongovett can reinvestigate at some point if this is actually caused by his machine or if this is a regression from the last few days.

> Is there anything special that I need to do to enable the SWC backend?

No, this looks correct.

**Reply by devongovett** [reply]
I'm on an M1. Maybe something is different there. One thing could be the new worker limiting logic which creates `cores / 2` threads. In the case of a 6 core machine rather than the 8 core M1, I think we might have determined 4 threads rather than 3 to be faster but I don't think we changed the logic yet. Anyway, there are many variables to benchmarking that make it difficult. Plus, Terser is the majority of this particular benchmark.

**Reply by evanw** [reply]
Interesting. Thanks for the context. I just updated my benchmark with the latest Parcel build but I'm happy to update it again when there's another build in the future. Unfortunately I don't have an M1 myself... I'm also looking forward to the Terser rewrite. That should dramatically improve most (all?) of the other bundlers in my benchmark.

**Reply by mischnic** [reply]
You should also be able to drop the tsc transformer for the Typescript benchmark. swc should handle it correctly (compared to Babel with that import shadowing bug)

**Reply by evanw** [reply]
That was my thought as well, but I tried that and unfortunately it didn't work. I can file a bug about it.

**Reply by evanw** [reply]
Here's the bug report: https://github.com/swc-project/swc/issues/1729.

**Reply by mischnic** [reply]
Thanks for investigating!
